### PR TITLE
[Runtime] Fixed several memory leaks

### PR DESCRIPTION
**Pull request by Dunati**
re-request for https://github.com/mono/mono/pull/3218

### CODE REVIEW

**Comment by akoeplinger** [comment]
@Dunati this causes crashes in almost all of the class library tests, e.g. https://jenkins.mono-project.com/job/test-mono-pull-request-amd64/3532/. Can you please take a look?

**Comment by akoeplinger** [comment]
Can you please squash this down to meaningful chunks and rebase out the merge commit? Thanks.

**Comment by steveisok** [comment]
@lambdageek should this be reviewed again or closed?

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -8504,6 +8503,14 @@ mono_icall_unlock (void)
 void
 mono_icall_cleanup (void)
 {
+	GHashTableIter iter;
+	gpointer key;
+
+	g_hash_table_iter_init (&iter, jit_icall_hash_name);
+	while (g_hash_table_iter_next (&iter, &key, NULL)) {
```
Comment (lewing) [inline mono/metadata/icall.c]: This isn't the right way to handle this g_hash_table_destroy will free the keys if you change the init call.

**Comment by lewing** [comment]
This isn't ready to merge and there hasn't been activity for a long time.  Closing it now

**Pull request closed without merging**